File: apps/web/lib/scoring.ts
export type Keypoint = [number, number, number];

const JOINT_WEIGHTS = [
  0.6, 0.3, 0.3, 0.2, 0.2, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.9, 0.9, 0.8, 0.8,
];

export function normalizePose(points: Keypoint[]): Keypoint[] {
  if (points.length !== 17) {
    return points;
  }
  const clone = points.map((item) => [...item] as Keypoint);
  const leftHip = clone[11];
  const rightHip = clone[12];
  const leftShoulder = clone[5];
  const rightShoulder = clone[6];

  const hipX = (leftHip[0] + rightHip[0]) / 2;
  const hipY = (leftHip[1] + rightHip[1]) / 2;
  const shoulderX = (leftShoulder[0] + rightShoulder[0]) / 2;
  const shoulderY = (leftShoulder[1] + rightShoulder[1]) / 2;
  const torso = Math.hypot(shoulderX - hipX, shoulderY - hipY) || 1;

  for (const item of clone) {
    item[0] = (item[0] - hipX) / torso;
    item[1] = (item[1] - hipY) / torso;
    item[2] = Math.max(0, Math.min(1, item[2]));
  }
  return clone;
}

export function weightedPoseSimilarity(player: Keypoint[], reference: Keypoint[]): number {
  if (player.length !== 17 || reference.length !== 17) {
    return 0;
  }
  let dot = 0;
  let pNorm = 0;
  let rNorm = 0;
  for (let i = 0; i < 17; i += 1) {
    const w = JOINT_WEIGHTS[i];
    const [px, py] = player[i];
    const [rx, ry] = reference[i];
    dot += w * (px * rx + py * ry);
    pNorm += w * (px * px + py * py);
    rNorm += w * (rx * rx + ry * ry);
  }
  if (pNorm < 1e-6 || rNorm < 1e-6) {
    return 0;
  }
  const cos = dot / Math.sqrt(pNorm * rNorm);
  const normalized = (cos + 1) / 2;
  return Math.max(0, Math.min(1, normalized));
}

export function mirrorPoseX(points: Keypoint[]): Keypoint[] {
  return points.map(([x, y, v]) => [-x, y, v]);
}

export function bestPoseSimilarity(player: Keypoint[], reference: Keypoint[]): number {
  const direct = weightedPoseSimilarity(player, reference);
  const mirrored = weightedPoseSimilarity(mirrorPoseX(player), reference);
  return Math.max(direct, mirrored);
}

export function panelScore(poseSim: number, offsetMs: number, windowMs: number): number {
  const timeSim = Math.max(0, 1 - Math.abs(offsetMs) / Math.max(1, windowMs));
  return Math.round(100 * (0.7 * poseSim + 0.3 * timeSim));
}

File: services/api/app/scoring.py
from __future__ import annotations

import math

import numpy as np

JOINT_WEIGHTS = np.array(
    [
        0.6,  # nose
        0.3,  # left_eye
        0.3,  # right_eye
        0.2,  # left_ear
        0.2,  # right_ear
        1.0,  # left_shoulder
        1.0,  # right_shoulder
        1.0,  # left_elbow
        1.0,  # right_elbow
        1.0,  # left_wrist
        1.0,  # right_wrist
        1.0,  # left_hip
        1.0,  # right_hip
        0.9,  # left_knee
        0.9,  # right_knee
        0.8,  # left_ankle
        0.8,  # right_ankle
    ],
    dtype=np.float32,
)


def weighted_pose_similarity(player: list[list[float]], reference: list[list[float]]) -> float:
    p = np.array(player, dtype=np.float32)
    r = np.array(reference, dtype=np.float32)
    if p.shape != (17, 3) or r.shape != (17, 3):
        return 0.0

    p_xy = p[:, :2].reshape(-1)
    r_xy = r[:, :2].reshape(-1)
    weights = np.repeat(JOINT_WEIGHTS, 2)
    p_weighted = p_xy * weights
    r_weighted = r_xy * weights

    p_norm = np.linalg.norm(p_weighted)
    r_norm = np.linalg.norm(r_weighted)
    if p_norm < 1e-6 or r_norm < 1e-6:
        return 0.0
    cosine = float(np.dot(p_weighted, r_weighted) / (p_norm * r_norm))
    return max(0.0, min(1.0, (cosine + 1.0) / 2.0))


def panel_score(pose_sim: float, offset_ms: int, window_ms: int) -> int:
    time_sim = max(0.0, 1.0 - (abs(offset_ms) / float(max(window_ms, 1))))
    score = round(100 * (0.7 * pose_sim + 0.3 * time_sim))
    return int(max(0, min(100, score)))


def rank_grade(final_score: int) -> str:
    if final_score >= 9000:
        return "S"
    if final_score >= 8000:
        return "A"
    if final_score >= 6500:
        return "B"
    if final_score >= 5000:
        return "C"
    return "D"

